lib/assert.js

```diff
@@ -333,3 +356,31 @@ assert.strict = Object.assign(strict, assert, {
   notDeepEqual: assert.notDeepStrictEqual
 });
 assert.strict.strict = assert.strict;
+
+// Asynchronous equivalent of original assert functions.
+// Only throws() and doNotThrows() are truely async (all other perform
+// synchronous tasks and don't use functions as parameters), but for
+// conveniency, the entire API is supported here as well
+
+function promises(...args) {
+  assert.ok(...args);
+}
+
+Object.assign(promises, assert, {
```

Comment:
Since other assigned APIs do not return promises here:

```js
require('assert').promises.fail().then(() => {}); // does not work
```

This behavior looks somewhat strange to me..

Reply:
To me as well. I would rather have a dedicated `assert.rejects` function. `Reject` is used for promises, so this is a good fit out of my perspective.

Reply:
That's an interesting point.

My intention was to keep the existing behaviour for all assert functions except `throws()` and `doesNotThrow()`, which are the only two able to run (potentially async) code.
I wouldn't like to force people writing:
```javascript
const assert = require('assert').promises

await assert(something)
```

But we can definitely add new functions like `reject()` that would be async equivalent of existing one. 
Are there any other function you would like to cover?

Reply:
If we are not talking about redesigning a new assert API here, providing a promisified version of the existing assert APIs is good enough to me. Something like

```js
for (const key of Object.keys(assert)) {
  promises[key] = async function (..args) { return assert[key](...args) };
}
```

Reply:
hum, but it's actually what I'd like to avoid.

There's no point making `assert.ok()` or `assert.fail()` asynchronous. It will make the API more cumbersome to use with no benefit.

It make sense for `assert.throws()` and `assert.doesNotThrow()` to turn this code (that we used quite a lot in our projects):
```javascript
it('should report database option errors', async () => {
  try {
    await transferData({}, '');
    throw new Error('it should have failed');
  } catch (err) {
    assert(err.message.includes('"host" is required'));
  }
});
```

into 
```javascript
it('should report database option errors', async () =>
  assert.throws(async () => transferData({}, ''), /^Error: "host" is required/)
);
```

I don't want to redesign the whole API, but rather wants to support the above.